### DISCUSSION TITLE
Post Fedora Linux upgrade/fresh installation to version 38 has caused paps to fail

### DISCUSSION THREAD

**Discussion by EPRINC-MP** [dov/paps#63] [General]
For the last ten or so years, I've relied on enscript and then paps. I think that I switched to paps because enscript at the time didn't support utf-8 (as best I can tell, it still doesn't).

paps has worked as designed for me until the upgrade/fresh install to Fedora 38. My sole availabe printer is an HP P3015. On my desktop, I use the HP Postscript driver; on my laptop I use the HP LaserJet p3010 Series hpijs pcl3, 3.22.10 driver for the printer.

When printing from a web browser (firefox of brave), libreoffice, evince, AcroRead, or gedit, the results are correct.

But now with the flip to F38, and issuing a paps command like this (which I've run for a decade):
paps --columns=3 --landscape --paper=letter --font="Monospace 6" /path/to/some/textfile | lpr -P lp3

the printer prints raw postscript.

(lp3 is the name that I use for the HP printer)

Could it be pango (required by paps)? some other possible dependency? Are there alternatives? Is there some nuance in configuring the printer that now needs to done with Fedora 38?

The latest paps version that is used on Fedora 37 is paps-0.7.1-5.fc37.x86_64

The latest paps version that is used on Fedora 38 is paps-0.7.9-1.fc38.x86_64

Much thanks for your help. I'm happy to help in testing to

**Comment by dov** [comment]
Glad you like paps. :-)

Do you mind doing the following two steps:

1. Instead of piping to lpr, redirect the output to a file, and then attach that file to this bug report.
2. Try compiling the latest version of paps 0.8.0 and see if it solves the problem, and if not attach the postscript output as well.

You may also try to set the format explicitly to posscript by the following flag `--format=ps` and see if it makes any difference.

**Reply by EPRINC-MP** [reply]
Thank you for your quick reply.

Request Number 1 - piping output to a file.

I created a simple file with one line "This is a test."

I did this on both the laptop and desktop and directed the output to TestOutput-paps.txt

I then compared the output on the two machines (in both cases the filesize is the same at 12572) and then ran sdiff
> sdiff -s TestOutput-paps-Laptop.txt TestOutput-paps-Desktop.txt 
%%CreationDate: Tue May  2 10:56:15 2023		      |	%%CreationDate: Tue May  2 10:55:13 2023
%%BeginResource: font #C2#B3#EF#BE#96#EE#BD#95		      |	%%BeginResource: font #C2#91#E6#81#8E#EC#8D#95
/FontName /#C2#B3#EF#BE#96#EE#BD#95 def			      |	/FontName /#C2#91#E6#81#8E#EC#8D#95 def

(sorry don't know how to format to monospace)
I'm attaching the output file

[TestOutput-paps-Laptop.txt](https://github.com/dov/paps/files/11374713/TestOutput-paps-Laptop.txt)

On to the following two requests.

M

**Reply by EPRINC-MP** [reply]
When compiling and running make (after satisfying dependencies for fmt-devel) I get the following output indicating failure of compilation:
> make
make  all-recursive
make[1]: Entering directory '/var/tmp/paps-0.8.0'
Making all in src
make[2]: Entering directory '/var/tmp/paps-0.8.0/src'
g++ -DHAVE_CONFIG_H -I. -I..  -DGETTEXT_PACKAGE='"paps"' -DDATADIR='"/usr/local/share"'  -fno-strict-aliasing -Wnested-externs -Wmissing-prototypes -Wstrict-prototypes -Wdeclaration-after-statement -Wimplicit-function-declaration -Wold-style-definition -Wjump-misses-init -Wall -Wextra -Wundef -Wwrite-strings -Wpointer-arith -Wmissing-declarations -Wredundant-decls -Wno-unused-parameter -Wno-missing-field-initializers -Wformat=2 -Wcast-align -Wformat-nonliteral -Wformat-security -Wsign-compare -Wstrict-aliasing -Wshadow -Winline -Wpacked -Wmissing-format-attribute -Wmissing-noreturn -Winit-self -Wmissing-include-dirs -Wunused-but-set-variable -Warray-bounds -Wreturn-type -Wswitch-enum -Wswitch-default -Wduplicated-cond -Wduplicated-branches -Wlogical-op -Wrestrict -Wnull-dereference -Wdouble-promotion -Werror -Wno-suggest-attribute=format -Wno-error=unused-parameter -Wno-error=missing-field-initializers -Wno-format-y2k -I/usr/include/pango-1.0 -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -I/usr/include/sysprof-4 -I/usr/include/harfbuzz -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/fribidi -I/usr/include/libxml2 -I/usr/include/cairo -I/usr/include/pixman-1 -pthread   -g -O2 -MT paps-paps.o -MD -MP -MF .deps/paps-paps.Tpo -c -o paps-paps.o `test -f 'paps.cc' || echo './'`paps.cc
cc1plus: warning: command-line option ‘-Wnested-externs’ is valid for C/ObjC but not for C++
cc1plus: warning: command-line option ‘-Wmissing-prototypes’ is valid for C/ObjC but not for C++
cc1plus: warning: command-line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++
cc1plus: warning: command-line option ‘-Wdeclaration-after-statement’ is valid for C/ObjC but not for C++
cc1plus: warning: command-line option ‘-Wimplicit-function-declaration’ is valid for C/ObjC but not for C++
cc1plus: warning: command-line option ‘-Wold-style-definition’ is valid for C/ObjC but not for C++
cc1plus: warning: command-line option ‘-Wjump-misses-init’ is valid for C/ObjC but not for C++
In file included from /usr/include/glib-2.0/glib/giochannel.h:36,
                 from /usr/include/glib-2.0/glib.h:56,
                 from /usr/include/glib-2.0/glib/gprintf.h:23,
                 from /usr/include/glib-2.0/glib/gstdio.h:24,
                 from paps.cc:23:
paps.cc: In function ‘bool copy_pango_parse_enum(GType, const char*, int*, bool, char**)’:
/usr/include/glib-2.0/glib/gstring.h:71:31: error: ignoring return value of ‘gchar* g_string_free_and_steal(GString*)’ declared with attribute ‘warn_unused_result’ [-Werror=unused-result]
   71 |       g_string_free_and_steal (str))            \
      |       ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~
paps.cc:388:11: note: in expansion of macro ‘g_string_free’
  388 |           g_string_free (s, possible_values ? false : true);
      |           ^~~~~~~~~~~~~
paps.cc: In function ‘char* read_file(FILE*, gchar*)’:
/usr/include/glib-2.0/glib/gstring.h:71:31: error: ignoring return value of ‘gchar* g_string_free_and_steal(GString*)’ declared with attribute ‘warn_unused_result’ [-Werror=unused-result]
   71 |       g_string_free_and_steal (str))            \
      |       ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~
paps.cc:1047:3: note: in expansion of macro ‘g_string_free’
 1047 |   g_string_free (inbuf, false);
      |   ^~~~~~~~~~~~~
paps.cc: In function ‘std::string get_date()’:
/usr/include/glib-2.0/glib/gstring.h:71:31: error: ignoring return value of ‘gchar* g_string_free_and_steal(GString*)’ declared with attribute ‘warn_unused_result’ [-Werror=unused-result]
   71 |       g_string_free_and_steal (str))            \
      |       ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~
paps.cc:1674:7: note: in expansion of macro ‘g_string_free’
 1674 |       g_string_free(inbuf, false);
      |       ^~~~~~~~~~~~~
/usr/include/glib-2.0/glib/gstring.h:71:31: error: ignoring return value of ‘gchar* g_string_free_and_steal(GString*)’ declared with attribute ‘warn_unused_result’ [-Werror=unused-result]
   71 |       g_string_free_and_steal (str))            \
      |       ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~
paps.cc:1683:5: note: in expansion of macro ‘g_string_free’
 1683 |     g_string_free(inbuf, false);
      |     ^~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
make[2]: *** [Makefile:454: paps-paps.o] Error 1
make[2]: Leaving directory '/var/tmp/paps-0.8.0/src'
make[1]: *** [Makefile:502: all-recursive] Error 1
make[1]: Leaving directory '/var/tmp/paps-0.8.0'
make: *** [Makefile:387: all] Error 2


Adivce? (Btw, yes, saw the output on my step 1)

**Comment by dov** [comment]
Found it. There is an initial debug line in the beginning of the output file that causes the magic number for postscript files to fail:

```
Got 1 pages
%!PS-Adobe-3.0
%%Creator...
```
This has been fixed in version 0.8.0 . Since this has been fixed upstream, in this repo, you should report this bug to Fedora. Please add a link to the Fedora bug to this issue for cross reference.

**Reply by EPRINC-MP** [reply]
Thank you. 

I've just reported this on the main Fedora discussion list. I don't know how to elevate this to appropriate personnel. 

Software used by small communities (such as paps, syncevolution, etc) have a way of being overlooked. 

But I will do my best to notify you.

M

**Reply by dov** [reply]
You probably want to report it to the Fedora Bugzilla and then add a link to the issue to this bug report.

Regarding the compilation, I don't know what is going on, and I haven't upgrade to Fedora 38 (yet). Note however that the main compilation environment for paps is meson so you may want to try that instead.

And finally, meanwhile, you mitigate the error by piping the paps output through the following one liner:

```
perl -ne 'print unless $.==1'
```

**Reply by EPRINC-MP** [reply]
This is in Fedora Rawhide (scheduled for the Fedora 39) release, but not in any F38 queue. 

I'm told to file a bugzilla report, which I've never done.
